pallets/contracts/src/lib.rs

```diff
@@ -0,0 +1,310 @@
+// This file is part of the Polymesh distribution (https://github.com/PolymathNetwork/Polymesh).
+// Copyright (c) 2020 Polymath
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, version 3.
+
+// This program is distributed in the hope that it will be useful, but
+// WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+// General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+// Ensure we're `no_std` when compiling for Wasm.
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use codec::Encode;
+use frame_support::{
+    decl_error, decl_event, decl_module, decl_storage,
+    dispatch::{DispatchError, DispatchResult, DispatchResultWithPostInfo},
+    ensure,
+    traits::Get,
+};
+use frame_system::{self as system, ensure_signed};
+use pallet_contracts::{BalanceOf, CodeHash, ContractAddressFor, Gas, Schedule};
+use pallet_identity as identity;
+use polymesh_common_utilities::{
+    identity::Trait as IdentityTrait,
+    protocol_fee::{ChargeProtocolFee, ProtocolOp},
+    Context,
+};
+use polymesh_primitives::{IdentityId, SmartExtensionMetadata, TemplateMetadata};
+use sp_core::crypto::UncheckedFrom;
+use sp_runtime::{
+    traits::{Hash, Saturating, StaticLookup},
+    Perbill, SaturatedConversion,
+};
+use sp_std::{marker::PhantomData, prelude::*};
+
+type Identity<T> = identity::Module<T>;
+
+/// Nonce based contract address determiner.
+///
+/// Address calculated from the code (of the constructor), input data to the constructor,
+/// the account id that requested the account creation and the nonce of the account id.
+///
+/// Formula: `blake2_256(blake2_256(code) + blake2_256(data) + origin + blake2_256(nonce))`
+pub struct NonceBasedAddressDeterminer<T: Trait>(PhantomData<T>);
+impl<T: Trait> ContractAddressFor<CodeHash<T>, T::AccountId> for NonceBasedAddressDeterminer<T>
+where
+    T::AccountId: UncheckedFrom<T::Hash> + AsRef<[u8]>,
+{
+    fn contract_address_for(
```

Comment:
`next_contract_address_for` makes more sense since you are calculating the address for the next nonce.

Reply:
It can't be changed as it is a function of the trait. 
You said next nonce it means for the current transaction `<frame_system::Module<T>>::account(origin).nonce` this will give a future nonce?

Reply:
I hadn't seen the nonce you are using. You should use a contracts module-specific nonce because account nonce is same across batch transactions.

Anyway, my main point was that this function returns the address for the latest nonce address. Therefore, the name `contract_address_for` is not technically correct. If you are sure this function is only used internally and only the latest nonce address is needed, please add a code comment about this behavior.